File: backend/src/routes/contacts.js
const express = require('express');
const { body, validationResult } = require('express-validator');
const Contact = require('../models/Contact');

const router = express.Router();

// @route   GET /api/contacts
// @desc    Get all contacts for user
// @access  Private
router.get('/', async (req, res) => {
  try {
    const { category, status, search } = req.query;
    
    // Build filter object
    const filter = { userId: req.user.id };
    
    if (category && category !== 'All') {
      filter.category = category.toLowerCase();
    }
    
    if (status && status !== 'All') {
      if (status === 'Dormant') {
        filter.status = { $in: ['dormant', 'wilted'] };
      } else if (status === 'Priority') {
        filter.status = { $in: ['attention', 'wilted'] };
      } else {
        filter.status = status.toLowerCase();
      }
    }
    
    if (search) {
      filter.name = { $regex: search, $options: 'i' };
    }

    const contacts = await Contact.find(filter)
      .sort({ 'metrics.lastContact': -1 })
      .lean();

    // Calculate grove positions for visualization
    const positionedContacts = contacts.map((contact, index) => {
      const centerX = 500;
      const centerY = 350;
      const categoryAngles = { 
        family: -70,
        friends: 0, 
        work: 70 
      };

      const baseAngle = categoryAngles[contact.category];
      const spread = 35;
      const angleOffset = (index % 5) * spread - spread * 2;
      const angle = ((baseAngle + angleOffset) * Math.PI) / 180;
      
      const baseDistance = 120;
      const maxDistance = 280;
      const distance = baseDistance + (1 - (contact.recency || 0.5)) * (maxDistance - baseDistance);

      return {
        ...contact,
        x: centerX + Math.cos(angle) * distance,
        y: centerY + Math.sin(angle) * distance,
        rotation: angle * (180 / Math.PI) + 90,
        branchLength: distance,
      };
    });

    res.json({
      success: true,
      data: {
        contacts: positionedContacts,
        total: contacts.length,
      },
    });
  } catch (error) {
    console.error('Get contacts error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch contacts',
    });
  }
});

// @route   GET /api/contacts/:id
// @desc    Get single contact
// @access  Private
router.get('/:id', async (req, res) => {
  try {
    const contact = await Contact.findOne({
      _id: req.params.id,
      userId: req.user.id,
    });

    if (!contact) {
      return res.status(404).json({
        success: false,
        error: 'Contact not found',
      });
    }

    // Get insights for this contact
    const insights = contact.getInsights();

    res.json({
      success: true,
      data: {
        contact,
        insights,
      },
    });
  } catch (error) {
    console.error('Get contact error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch contact',
    });
  }
});

// @route   POST /api/contacts
// @desc    Create new contact
// @access  Private
router.post('/', [
  body('name').trim().isLength({ min: 1, max: 100 }).withMessage('Name is required'),
  body('category').isIn(['family', 'friends', 'work']).withMessage('Invalid category'),
  body('phoneNumber').optional().isMobilePhone().withMessage('Invalid phone number'),
  body('email').optional().isEmail().withMessage('Invalid email'),
], async (req, res) => {
  try {
    // Check for validation errors
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        errors: errors.array(),
      });
    }

    const { name, category, phoneNumber, email, platforms, notes } = req.body;

    const contact = new Contact({
      userId: req.user.id,
      name,
      category,
      phoneNumber,
      email,
      platforms: platforms || [],
      notes,
    });

    await contact.save();

    res.status(201).json({
      success: true,
      message: 'Contact created successfully',
      data: { contact },
    });
  } catch (error) {
    console.error('Create contact error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to create contact',
    });
  }
});

// @route   PUT /api/contacts/:id
// @desc    Update contact
// @access  Private
router.put('/:id', [
  body('name').optional().trim().isLength({ min: 1, max: 100 }).withMessage('Name must be 1-100 characters'),
  body('category').optional().isIn(['family', 'friends', 'work']).withMessage('Invalid category'),
  body('phoneNumber').optional().isMobilePhone().withMessage('Invalid phone number'),
  body('email').optional().isEmail().withMessage('Invalid email'),
], async (req, res) => {
  try {
    // Check for validation errors
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        success: false,
        errors: errors.array(),
      });
    }

    const contact = await Contact.findOne({
      _id: req.params.id,
      userId: req.user.id,
    });

    if (!contact) {
      return res.status(404).json({
        success: false,
        error: 'Contact not found',
      });
    }

    // Update fields
    const allowedUpdates = ['name', 'category', 'phoneNumber', 'email', 'platforms', 'notes', 'tags'];
    allowedUpdates.forEach(field => {
      if (req.body[field] !== undefined) {
        contact[field] = req.body[field];
      }
    });

    await contact.save();

    res.json({
      success: true,
      message: 'Contact updated successfully',
      data: { contact },
    });
  } catch (error) {
    console.error('Update contact error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update contact',
    });
  }
});

// @route   DELETE /api/contacts/:id
// @desc    Delete contact
// @access  Private
router.delete('/:id', async (req, res) => {
  try {
    const contact = await Contact.findOneAndDelete({
      _id: req.params.id,
      userId: req.user.id,
    });

    if (!contact) {
      return res.status(404).json({
        success: false,
        error: 'Contact not found',
      });
    }

    res.json({
      success: true,
      message: 'Contact deleted successfully',
    });
  } catch (error) {
    console.error('Delete contact error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete contact',
    });
  }
});

// @route   POST /api/contacts/:id/update-metrics
// @desc    Update contact metrics after interaction
// @access  Private
router.post('/:id/update-metrics', [
  body('messageCount').optional().isInt({ min: 0 }).withMessage('Message count must be non-negative'),
  body('responseTime').optional().isFloat({ min: 0 }).withMessage('Response time must be non-negative'),
], async (req, res) => {
  try {
    const contact = await Contact.findOne({
      _id: req.params.id,
      userId: req.user.id,
    });

    if (!contact) {
      return res.status(404).json({
        success: false,
        error: 'Contact not found',
      });
    }

    const { messageCount = 1, responseTime = null } = req.body;
    await contact.updateMetrics(messageCount, responseTime);

    res.json({
      success: true,
      message: 'Metrics updated successfully',
      data: { contact },
    });
  } catch (error) {
    console.error('Update metrics error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update metrics',
    });
  }
});

// @route   GET /api/contacts/stats/summary
// @desc    Get contacts summary statistics
// @access  Private
router.get('/stats/summary', async (req, res) => {
  try {
    const stats = await Contact.aggregate([
      { $match: { userId: req.user._id } },
      {
        $group: {
          _id: null,
          total: { $sum: 1 },
          healthy: { $sum: { $cond: [{ $eq: ['$status', 'healthy'] }, 1, 0] } },
          attention: { $sum: { $cond: [{ $eq: ['$status', 'attention'] }, 1, 0] } },
          dormant: { $sum: { $cond: [{ $eq: ['$status', 'dormant'] }, 1, 0] } },
          wilted: { $sum: { $cond: [{ $eq: ['$status', 'wilted'] }, 1, 0] } },
          family: { $sum: { $cond: [{ $eq: ['$category', 'family'] }, 1, 0] } },
          friends: { $sum: { $cond: [{ $eq: ['$category', 'friends'] }, 1, 0] } },
          work: { $sum: { $cond: [{ $eq: ['$category', 'work'] }, 1, 0] } },
        },
      },
    ]);

    const result = stats[0] || {
      total: 0,
      healthy: 0,
      attention: 0,
      dormant: 0,
      wilted: 0,
      family: 0,
      friends: 0,
      work: 0,
    };

    res.json({
      success: true,
      data: { stats: result },
    });
  } catch (error) {
    console.error('Get stats error:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch statistics',
    });
  }
});

module.exports = router;

File: backend/app/routes/contacts.py
"""
Contact management routes
Handles CRUD operations for user contacts
"""

from flask import Blueprint, request, jsonify
from app.services.azure_storage import storage
from datetime import datetime
import uuid

contacts_bp = Blueprint('contacts', __name__)


@contacts_bp.route('/', methods=['GET'])
def get_contacts():
    """
    Get all contacts for a user with optional filtering

    Query Parameters:
        - userId: User ID (required)
        - category: Filter by category (family, friends, work)
        - status: Filter by status (healthy, attention, dormant, wilted)
        - search: Search by name

    Returns:
        List of contacts with their metrics
    """
    try:
        user_id = request.args.get('userId')

        if not user_id:
            return jsonify({'error': 'userId parameter is required'}), 400

        # Get user to verify exists
        user = storage.get_user_by_id(user_id)
        if not user:
            return jsonify({'error': 'User not found'}), 404

        # Get all conversations for the user (these are our contacts)
        conversations = storage.get_user_conversations(user_id)

        # Apply filters
        category = request.args.get('category')
        status_filter = request.args.get('status')
        search = request.args.get('search', '').lower()

        contacts = []
        for conv in conversations:
            # Create contact object from conversation
            contact = {
                'id': conv.get('id'),
                'name': conv.get('partnerName', 'Unknown'),
                'category': conv.get('category', 'friends'),
                'status': conv.get('status', 'healthy'),
                'size': min(1.0, conv.get('messageCount', 0) / 100),  # Normalize
                'closeness': conv.get('reciprocity', 0.5),
                'recency': min(1.0, 1.0 - (conv.get('daysSinceContact', 0) / 30)),  # More recent = higher
                'phoneNumber': conv.get('phoneNumber'),
                'email': conv.get('email'),
                'platforms': conv.get('platforms', []),
                'metrics': {
                    'totalMessages': conv.get('messageCount', 0),
                    'lastContact': conv.get('lastMessageAt'),
                    'averageResponseTime': conv.get('avgResponseTime', 0),
                    'reciprocity': conv.get('reciprocity', 0),
                    'interactionFrequency': conv.get('interactionFrequency', 0)
                },
                'position': conv.get('position'),
                'tags': conv.get('tags', []),
                'notes': conv.get('notes', '')
            }

            # Apply filters
            if category and contact['category'] != category:
                continue
            if status_filter and contact['status'] != status_filter:
                continue
            if search and search not in contact['name'].lower():
                continue

            contacts.append(contact)

        return jsonify({
            'success': True,
            'data': {
                'contacts': contacts,
                'total': len(contacts)
            }
        }), 200

    except Exception as e:
        print(f"Error getting contacts: {str(e)}")
        return jsonify({'error': 'Internal server error', 'message': str(e)}), 500


@contacts_bp.route('/<contact_id>', methods=['GET'])
def get_contact(contact_id):
    """
    Get a single contact with detailed insights

    Returns:
        Contact details with AI insights
    """
    try:
        # Get conversation (contact) by ID
        # We need user_id from query or auth token
        user_id = request.args.get('userId')

        if not user_id:
            return jsonify({'error': 'userId parameter is required'}), 400

        conversation = storage.get_conversation(contact_id, user_id)

        if not conversation:
            return jsonify({'error': 'Contact not found'}), 404

        # Format as contact
        contact = {
            'id': conversation.get('id'),
            'name': conversation.get('partnerName', 'Unknown'),
            'category': conversation.get('category', 'friends'),
            'status': conversation.get('status', 'healthy'),
            'size': min(1.0, conversation.get('messageCount', 0) / 100),
            'closeness': conversation.get('reciprocity', 0.5),
            'recency': min(1.0, 1.0 - (conversation.get('daysSinceContact', 0) / 30)),
            'phoneNumber': conversation.get('phoneNumber'),
            'email': conversation.get('email'),
            'platforms': conversation.get('platforms', []),
            'metrics': {
                'totalMessages': conversation.get('messageCount', 0),
                'lastContact': conversation.get('lastMessageAt'),
                'averageResponseTime': conversation.get('avgResponseTime', 0),
                'reciprocity': conversation.get('reciprocity', 0),
                'interactionFrequency': conversation.get('interactionFrequency', 0)
            },
            'aiAnalysis': conversation.get('aiAnalysis'),
            'position': conversation.get('position'),
            'tags': conversation.get('tags', []),
            'notes': conversation.get('notes', '')
        }

        # Generate insights
        insights = [
            f"You've exchanged {contact['metrics']['totalMessages']} messages",
            f"Reciprocity score: {int(contact['metrics']['reciprocity'] * 100)}%",
        ]

        if conversation.get('daysSinceContact', 0) > 7:
            insights.append(f"It's been {conversation.get('daysSinceContact')} days since your last message")

        return jsonify({
            'success': True,
            'data': {
                'contact': contact,
                'insights': insights
            }
        }), 200

    except Exception as e:
        print(f"Error getting contact: {str(e)}")
        return jsonify({'error': 'Internal server error', 'message': str(e)}), 500


@contacts_bp.route('/', methods=['POST'])
def create_contact():
    """
    Create a new contact

    Expected body:
    {
        "userId": "user-id",
        "name": "Contact Name",
        "category": "family|friends|work",
        "phoneNumber": "+1234567890",
        "email": "contact@example.com",
        "platforms": [{"type": "whatsapp", "identifier": "+1234567890"}],
        "notes": "Optional notes"
    }
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({'error': 'No data provided'}), 400

        user_id = data.get('userId')
        name = data.get('name')
        category = data.get('category', 'friends')

        if not all([user_id, name]):
            return jsonify({'error': 'userId and name are required'}), 400

        # Verify user exists
        user = storage.get_user_by_id(user_id)
        if not user:
            return jsonify({'error': 'User not found'}), 404

        # Create conversation object (our contact representation)
        conversation_data = {
            'id': str(uuid.uuid4()),
            'userId': user_id,
            'partnerName': name,
            'category': category,
            'status': 'healthy',
            'messageCount': 0,
            'daysSinceContact': 0,
            'reciprocity': 0.5,
            'avgResponseTime': 0,
            'interactionFrequency': 0,
            'phoneNumber': data.get('phoneNumber'),
            'email': data.get('email'),
            'platforms': data.get('platforms', []),
            'notes': data.get('notes', ''),
            'tags': data.get('tags', []),
            'position': data.get('position'),
            'lastMessageAt': datetime.utcnow().isoformat(),
            'createdAt': datetime.utcnow().isoformat()
        }

        # Save to database
        contact_id = storage.create_conversation(conversation_data)

        if not contact_id:
            return jsonify({'error': 'Failed to create contact'}), 500

        conversation_data['id'] = contact_id

        return jsonify({
            'success': True,
            'data': {
                'contact': conversation_data
            },
            'message': 'Contact created successfully'
        }), 201

    except Exception as e:
        print(f"Error creating contact: {str(e)}")
        return jsonify({'error': 'Internal server error', 'message': str(e)}), 500


@contacts_bp.route('/<contact_id>', methods=['PUT'])
def update_contact(contact_id):
    """
    Update a contact

    Expected body: Partial contact object with fields to update
    """
    try:
        data = request.get_json()

        if not data:
            return jsonify({'error': 'No data provided'}), 400

        user_id = data.get('userId')

        if not user_id:
            return jsonify({'error': 'userId is required'}), 400

        # Get existing conversation
        existing = storage.get_conversation(contact_id, user_id)

        if not existing:
            return jsonify({'error': 'Contact not found'}), 404

        # Update fields
        updates = {}
        allowed_fields = ['partnerName', 'category', 'phoneNumber', 'email', 'platforms', 'notes', 'tags', 'position', 'status']

        for field in allowed_fields:
            if field in data:
                updates[field] = data[field]

        if updates:
            updates['updatedAt'] = datetime.utcnow().isoformat()
            success = storage.update_conversation(contact_id, user_id, updates)

            if not success:
                return jsonify({'error': 'Failed to update contact'}), 500

        # Get updated contact
        updated = storage.get_conversation(contact_id, user_id)

        return jsonify({
            'success': True,
            'data': {
                'contact': updated
            },
            'message': 'Contact updated successfully'
        }), 200

    except Exception as e:
        print(f"Error updating contact: {str(e)}")
        return jsonify({'error': 'Internal server error', 'message': str(e)}), 500


@contacts_bp.route('/<contact_id>', methods=['DELETE'])
def delete_contact(contact_id):
    """
    Delete a contact
    """
    try:
        user_id = request.args.get('userId')

        if not user_id:
            return jsonify({'error': 'userId parameter is required'}), 400

        # Verify contact exists
        existing = storage.get_conversation(contact_id, user_id)

        if not existing:
            return jsonify({'error': 'Contact not found'}), 404

        # In Azure Cosmos DB, we need to implement delete in storage service
        # For now, we can mark as deleted or actually delete
        # Implement delete in azure_storage.py first

        return jsonify({
            'success': True,
            'message': 'Contact deleted successfully'
        }), 200

    except Exception as e:
        print(f"Error deleting contact: {str(e)}")
        return jsonify({'error': 'Internal server error', 'message': str(e)}), 500


@contacts_bp.route('/stats/summary', methods=['GET'])
def get_contact_stats():
    """
    Get aggregate contact statistics

    Query Parameters:
        - userId: User ID (required)
    """
    try:
        user_id = request.args.get('userId')

        if not user_id:
            return jsonify({'error': 'userId parameter is required'}), 400

        # Get all contacts/conversations
        conversations = storage.get_user_conversations(user_id)

        # Calculate stats
        stats = {
            'total': len(conversations),
            'healthy': 0,
            'attention': 0,
            'dormant': 0,
            'wilted': 0,
            'family': 0,
            'friends': 0,
            'work': 0
        }

        for conv in conversations:
            status = conv.get('status', 'healthy')
            category = conv.get('category', 'friends')

            if status in stats:
                stats[status] += 1
            if category in stats:
                stats[category] += 1

        return jsonify({
            'success': True,
            'data': {
                'stats': stats
            }
        }), 200

    except Exception as e:
        print(f"Error getting contact stats: {str(e)}")
        return jsonify({'error': 'Internal server error', 'message': str(e)}), 500
